src/main/java/com/aws/greengrass/clientdevices/auth/ClientDevicesAuthService.java

```diff
@@ -198,21 +219,81 @@ protected void install() throws InterruptedException {
+
             if (whatHappened == WhatHappened.initialized || node == null) {
                 updateDeviceGroups(whatHappened, deviceGroupTopics);
-                updateCAType(caTypeTopic);
+                updateCertificateAuthority(caType, caCertPath, caKeyPath);
             } else if (node.childOf(DEVICE_GROUPS_TOPICS)) {
                 updateDeviceGroups(whatHappened, deviceGroupTopics);
-            } else if (node.childOf(CA_TYPE_TOPIC)) {
-                if (caTypeTopic.getOnce() == null) {
+            } else if (node.childOf(CERTIFICATE_AUTHORITY_TOPIC)) {
+                // If the CA type is null because of the change to its configuration to null, do not update the CA.
+
+                // If the CA type is null and if either of the customer provided CA key and CA certificate (or both)
+                // are empty, do not update the CA.
+                if (caType.getOnce() == null
+                        && (node.childOf(CA_TYPE_TOPIC) || Utils.isEmpty(caKeyPath) || Utils.isEmpty(caCertPath))) {
                     return;
                 }
-                updateCAType(caTypeTopic);
+                updateCertificateAuthority(caType, caCertPath, caKeyPath);
             }
         });
 
     }
 
+    private void updateCertificateAuthority(Topic caType, String caCertPath, String caKeyPath) {
+        // If the key or cert path of the CA are empty, generate a CA
+        if (Utils.isEmpty(caKeyPath) || Utils.isEmpty(caCertPath)) {
+            updateCAType(caType);
+        } else {
+            logger.atInfo().kv(CA_PRIVATE_KEY_URI, caKeyPath).kv(CA_CERTIFICATE_URI, caCertPath)
+                    .log("Using certificate authority URIs provided in the configuration.");
+            // Update certificate manager with the customer provided CA configuration
+            try {
+                URI privateKeyUri = new URI(caKeyPath);
+                URI certificateUri = new URI(caCertPath);
+                updateCertificateManager(privateKeyUri, certificateUri);
+            } catch (URISyntaxException e) {
+                logger.atError().kv(CA_PRIVATE_KEY_URI, caKeyPath).kv(CA_CERTIFICATE_URI, caCertPath).setCause(e)
+                        .log("Exception occurred while parsing the certificate authority configuration");
+                serviceErrored(e);
+            }
+        }
+        // Upload the generated or provided CA certificates to the GG cloud and update config
+        try {
+            List<String> caCerts = certificateManager.getCACertificates();
+            uploadCoreDeviceCAs(caCerts);
+            updateCACertificateConfig(caCerts);
+        } catch (KeyStoreException | IOException | CertificateEncodingException e) {
+            logger.atError().cause(e).log("Failed to get the CA certificates");
+            serviceErrored(e);
+        }
+    }
+
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
+    private void updateCertificateManager(URI privateKeyUri, URI certificateUri) {
```

Comment:
what thread is running this?

If this is running in the main context thread, this will not work. The service unavailable exception is the only exception which should be retried.

The service will only become available when it registers itself, and to do that it will be constructed by the context thread. So make sure we are not in the context thread, otherwise this is guaranteed to fail.